File: ios/Sources/ChoisyVideoKit/Thumbnailer.swift
@preconcurrency import AVFoundation
import CoreGraphics
import CryptoKit
import Foundation
import ImageIO
import UniformTypeIdentifiers

/// What `probe({ uri })` resolves with. All six keys are always present: `rotation` is informational
/// but the TS contract declares it, and an absent key reads as `undefined` in JS, which is a
/// different thing from `0`.
struct ProbeResult: Sendable {
    let durationMs: Int64
    /// DISPLAY dimensions, with the track's preferred transform already applied, so a portrait
    /// recording that stores 1920x1080 plus a 90 degree rotation reports 1080x1920 the way the
    /// editor draws it.
    let width: Int
    let height: Int
    let rotation: Int
    let hasAudio: Bool
    let hasVideo: Bool

    var json: [String: Any] {
        ["durationMs": durationMs, "width": width, "height": height,
         "rotation": rotation, "hasAudio": hasAudio, "hasVideo": hasVideo]
    }
}

/// Frame extraction and file inspection, shared by the plugin's `probe` and `thumbnails` methods,
/// by the exporter's result builder and by the voice recorder.
///
/// Android reaches for `MediaMetadataRetriever` for all three jobs; iOS splits them between
/// `AVURLAsset` property loads and `AVAssetImageGenerator`, but the answers are deliberately the
/// same shape so a filmstrip cut on one platform lines up with the other's.
enum Thumbnailer {

    // MARK: - Probe

    /// Works on audio-only files, answering `hasVideo: false` with zero dimensions: the editor
    /// probes music and voice takes through this same call purely for their duration.
    ///
    /// Every caller treats `durationMs <= 0` as "could not read it" and falls back, so returning 0
    /// is a legitimate answer. Throwing is reserved for a file that cannot be opened at all.
    static func probe(_ url: URL) async throws -> ProbeResult {
        // Precise timing costs a container scan on a file with no duration atom, which is exactly
        // the case (a still-muxing recording) where the cheap answer is wrong.
        let asset = AVURLAsset(url: url, options: [AVURLAssetPreferPreciseDurationAndTimingKey: true])
        let duration = try await asset.load(.duration)
        let video = try await asset.loadTracks(withMediaType: .video).first
        let audio = try await asset.loadTracks(withMediaType: .audio).first

        var width = 0
        var height = 0
        var rotation = 0
        if let track = video {
            let (natural, transform) = try await track.load(.naturalSize, .preferredTransform)
            // Transform the RECT, not the size. CGSize.applying can hand back negative components
            // for a 90 or 180 degree transform, while a rect's bounding box is always positive.
            let box = CGRect(origin: .zero, size: natural).applying(transform)
            width = Int(abs(box.width).rounded())
            height = Int(abs(box.height).rounded())
            rotation = rotationDegrees(transform)
        }

        // An indefinite or invalid duration (a live stream, a truncated container) makes `.seconds`
        // NaN, and Int64(NaN) traps rather than returning anything.
        let seconds = duration.seconds
        let durationMs = (duration.isNumeric && seconds.isFinite) ? Int64((seconds * 1000).rounded()) : 0

        return ProbeResult(durationMs: max(0, durationMs), width: width, height: height,
                           rotation: rotation, hasAudio: audio != nil, hasVideo: video != nil)
    }

    /// Clockwise degrees, the same convention as Android's `METADATA_KEY_VIDEO_ROTATION`:
    /// identity is 0, (0,1,-1,0) is 90, (-1,0,0,-1) is 180, (0,-1,1,0) is 270.
    ///
    /// The snap to a quarter turn is not cosmetic: a mirrored front-camera transform carries a
    /// scale of -1 on one axis and lands a degree or two off, and the contract promises one of
    /// exactly four values.
    static func rotationDegrees(_ t: CGAffineTransform) -> Int {
        let radians = atan2(t.b, t.a)
        guard radians.isFinite else { return 0 }
        var degrees = Int((radians * 180 / .pi).rounded())
        degrees = ((degrees % 360) + 360) % 360
        return Int((Double(degrees) / 90).rounded()) * 90 % 360
    }

    // MARK: - Filmstrip

    /// Returns one URL per requested time, in the requested order, always. The editor indexes the
    /// returned array against its own time array to position tiles, so a dropped entry does not
    /// shorten the strip, it shifts every tile after it against the timeline.
    ///
    /// Cached on disk by (source signature, time, maxHeight, precise). Reopening a clip in the
    /// editor is the common case and it should not decode a single frame.
    static func thumbnails(_ url: URL, timesMs: [Int64], maxHeight rawMaxHeight: Int,
                           precise: Bool) async throws -> [URL] {
        if timesMs.isEmpty { return [] }                            // never open a decoder for nothing
        let maxHeight = min(1080, max(16, rawMaxHeight))            // Android's coerceIn(16, 1080)
        let dir = JobFolders.thumbsDir()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let sourceKey = cacheKey(for: url)

        // Pass 1: whatever is already on disk. Android checks this per time before it opens a
        // retriever, and the editor's own "paid once, the frames are cached by time" comment
        // depends on it being here rather than in JS.
        var results = [URL?](repeating: nil, count: timesMs.count)
        var wanted: [(index: Int, timeMs: Int64)] = []
        for (i, raw) in timesMs.enumerated() {
            let clamped = max(0, raw)
            let file = dir.appendingPathComponent(cacheName(sourceKey, clamped, maxHeight, precise))
            if fileBytes(file) > 0 {
                results[i] = file
            } else {
                wanted.append((i, clamped))
            }
        }

        if !wanted.isEmpty {
            // A duplicated time (the strip's step can land twice on a very short clip) must not be
            // asked for twice; the generator would decode it twice for one picture.
            var uniqueTimes: [Int64] = []
            var seen = Set<Int64>()
            for entry in wanted where !seen.contains(entry.timeMs) {
                seen.insert(entry.timeMs)
                uniqueTimes.append(entry.timeMs)
            }
            uniqueTimes.sort()

            let generator = AVAssetImageGenerator(asset: AVURLAsset(url: url))
            generator.appliesPreferredTrackTransform = true         // a portrait clip comes out portrait
            // Android's getScaledFrameAtTime box. A full box rather than an unconstrained width,
            // so an ultra-wide source produces the same tile on both platforms.
            generator.maximumSize = CGSize(width: maxHeight * 2, height: maxHeight)
            // `precise` maps to tolerances and to nothing else. false is Android's
            // OPTION_CLOSEST_SYNC (jump to the keyframe, one decode), true is OPTION_CLOSEST
            // (decode forward from the keyframe before the time). A fixed half-second tolerance
            // would be neither, and the two platforms would disagree about which frame a tile shows.
            generator.requestedTimeToleranceBefore = precise ? .zero : .positiveInfinity
            generator.requestedTimeToleranceAfter = precise ? .zero : .positiveInfinity

            // Key by the REQUESTED time rather than trusting the sequence order: the batch mode is
            // documented as time ordered, not request ordered, and we asked in ascending order.
            var byTime: [Int64: URL] = [:]
            for await element in generator.images(for: uniqueTimes.map { ms($0) }) {
                guard case .success(requestedTime: let time, image: let image, actualTime: _) = element else {
                    continue                                        // fillGaps deals with it below
                }
                let key = msOf(time)
                let file = dir.appendingPathComponent(cacheName(sourceKey, key, maxHeight, precise))
                if let written = try? writeJPEG(downsample(image, maxHeight: maxHeight), to: file, quality: 0.8) {
                    byTime[key] = written
                }
            }
            for entry in wanted { results[entry.index] = byTime[entry.timeMs] }
        }

        return fillGaps(results, in: dir)
    }

    /// `"<sourceKey>-<timeMs>-<maxHeight>[-p].jpg"`. The height and the precise suffix both belong
    /// in the name: a precise tile is a different picture of the same moment, and a strip asked for
    /// precisely must never be served the keyframes an earlier sloppy request left behind.
    private static func cacheName(_ sourceKey: String, _ timeMs: Int64, _ maxHeight: Int, _ precise: Bool) -> String {
        "\(sourceKey)-\(timeMs)-\(maxHeight)\(precise ? "-p" : "").jpg"
    }

    /// The signature covers path, length and modification time, exactly as Android's does, so a
    /// re-recorded take written to the same job-folder path can never serve the old clip's strip.
    private static func cacheKey(for url: URL) -> String {
        let values = try? url.resourceValues(forKeys: [.fileSizeKey, .contentModificationDateKey])
        let size = values?.fileSize ?? 0
        let modified = Int64(((values?.contentModificationDate?.timeIntervalSince1970 ?? 0) * 1000).rounded())
        let signature = "\(url.absoluteString)|\(size)|\(modified)"
        let digest = SHA256.hash(data: Data(signature.utf8))
        return String(digest.map { String(format: "%02x", $0) }.joined().prefix(24))
    }

    /// Android's `fillGaps`: nearest earlier tile, then nearest later one, then one shared
    /// placeholder. Filling forward from an already-filled slot is intentional and matches Android;
    /// both routes end at the same picture.
    private static func fillGaps(_ urls: [URL?], in dir: URL) -> [URL] {
        guard urls.contains(where: { $0 == nil }) else { return urls.compactMap { $0 } }
        var filled = urls
        for i in filled.indices where filled[i] == nil {
            filled[i] = (0..<i).reversed().compactMap { filled[$0] }.first
                ?? ((i + 1)..<filled.count).compactMap { filled[$0] }.first
        }
        guard filled.contains(where: { $0 == nil }) else { return filled.compactMap { $0 } }
        let placeholder = blackTile(in: dir)
        return filled.map { $0 ?? placeholder }
    }

    /// One shared 2x2 black tile, written once under a name that cannot collide with a real time.
    ///
    /// Never write this under a requested time's cache name. The cache-hit pass above would serve
    /// it forever, so a single transient decode failure would black that tile out for the life of
    /// the file.
    private static func blackTile(in dir: URL) -> URL {
        let file = dir.appendingPathComponent("placeholder.jpg")
        if fileBytes(file) > 0 { return file }
        if let ctx = CGContext(data: nil, width: 2, height: 2, bitsPerComponent: 8, bytesPerRow: 0,
                               space: CGColorSpaceCreateDeviceRGB(),
                               bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue) {
            ctx.setFillColor(red: 0, green: 0, blue: 0, alpha: 1)
            ctx.fill(CGRect(x: 0, y: 0, width: 2, height: 2))
            if let image = ctx.makeImage() { _ = try? writeJPEG(image, to: file, quality: 0.8) }
        }
        return file
    }

    // MARK: - Poster

    /// Android's `OPTION_CLOSEST` then `OPTION_CLOSEST_SYNC`, translated: ask for the exact frame
    /// first, fall back to the nearest keyframe.
    ///
    /// Returns false rather than throwing. A finished render is never failed because its poster
    /// could not be cut; the caller sends `posterUri: ""` and the server cuts its own.
    static func poster(from videoURL: URL, atMs: Int64, to dest: URL) async -> Bool {
        let generator = AVAssetImageGenerator(asset: AVURLAsset(url: videoURL))
        generator.appliesPreferredTrackTransform = true
        generator.maximumSize = .zero                               // .zero is "do not scale"
        // Our own output carries roughly one keyframe a second, so half a second off is a visibly
        // different moment. Ask precisely first.
        generator.requestedTimeToleranceBefore = .zero
        generator.requestedTimeToleranceAfter = .zero

        let at = ms(max(0, atMs))
        try? FileManager.default.createDirectory(at: dest.deletingLastPathComponent(),
                                                 withIntermediateDirectories: true)
        if let image = try? await generator.image(at: at).image,
           (try? writeJPEG(image, to: dest, quality: 0.85)) != nil {
            return true
        }
        // The second attempt has to be a DIFFERENT request. Retrying at zero tolerance repeats the
        // one that just failed; infinite tolerance is what lets the decoder answer with the
        // keyframe it already has.
        generator.requestedTimeToleranceBefore = .positiveInfinity
        generator.requestedTimeToleranceAfter = .positiveInfinity
        if let image = try? await generator.image(at: at).image,
           (try? writeJPEG(image, to: dest, quality: 0.85)) != nil {
            return true
        }
        return false
    }

    // MARK: - Pixels and bytes

    /// Belt and braces around `maximumSize`, which is a bounding box and rounds to the encoder's
    /// liking: the contract promises the longest edge of the produced JPEG, not an approximation.
    private static func downsample(_ image: CGImage, maxHeight: Int) -> CGImage {
        guard image.height > maxHeight, image.height > 0 else { return image }
        let scale = Double(maxHeight) / Double(image.height)
        let width = max(1, Int((Double(image.width) * scale).rounded()))
        guard let ctx = CGContext(data: nil, width: width, height: maxHeight, bitsPerComponent: 8,
                                  bytesPerRow: 0, space: CGColorSpaceCreateDeviceRGB(),
                                  bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue) else { return image }
        ctx.interpolationQuality = .high
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: width, height: maxHeight))
        return ctx.makeImage() ?? image
    }

    /// ImageIO rather than UIKit, so none of this needs a main thread or a UIImage round trip.
    ///
    /// The file is written under a hidden sibling name and moved into place, because the cache-hit
    /// check is a size test: a half-written file that an interrupted call left behind would look
    /// like a valid tile forever.
    @discardableResult
    private static func writeJPEG(_ image: CGImage, to dest: URL, quality: Double) throws -> URL {
        let tmp = dest.deletingLastPathComponent().appendingPathComponent(".\(UUID().uuidString).jpg")
        guard let out = CGImageDestinationCreateWithURL(tmp as CFURL, UTType.jpeg.identifier as CFString, 1, nil) else {
            throw CocoaError(.fileWriteUnknown)
        }
        CGImageDestinationAddImage(out, image, [kCGImageDestinationLossyCompressionQuality: quality] as CFDictionary)
        guard CGImageDestinationFinalize(out) else {
            try? FileManager.default.removeItem(at: tmp)
            throw CocoaError(.fileWriteUnknown)
        }
        try? FileManager.default.removeItem(at: dest)               // moveItem refuses an existing destination
        do {
            try FileManager.default.moveItem(at: tmp, to: dest)
        } catch {
            try? FileManager.default.removeItem(at: tmp)
            throw error
        }
        return dest
    }

    /// 0 for a file that is not there, which is what every caller means by "unusable".
    static func fileBytes(_ url: URL) -> Int64 {
        guard let size = try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize else { return 0 }
        return Int64(size)
    }
}

File: android/src/main/java/net/dotnetdreamer/videokit/videocomposer/Thumbnailer.kt
package net.dotnetdreamer.videokit.videocomposer

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import android.provider.OpenableColumns
import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest

/**
 * Everything that needs to look inside a media file without rendering it: durations, track
 * presence, filmstrip frames and the poster cut from a finished render.
 *
 * `MediaMetadataRetriever` opens a decoder, and decoders are a scarce, device-limited resource that
 * the preview and the feed are also competing for. So every batch shares ONE retriever, batches for
 * the same source are serialised, and frames are cached on disk - a filmstrip is asked for again
 * every time the editor reopens.
 */
object Thumbnailer {

    private const val TAG = "VideoComposer"
    private const val JPEG_QUALITY_THUMB = 80
    private const val JPEG_QUALITY_POSTER = 85

    /** One lock per source, so two filmstrip requests for one file do not open two decoders. */
    private val locks = HashMap<String, Any>()

    private fun lockFor(key: String): Any = synchronized(locks) { locks.getOrPut(key) { Any() } }

    /* ---------------------------------------------------------------------------------------- */

    fun probe(ctx: Context, uri: String): ProbeInfo {
        val retriever = MediaMetadataRetriever()
        try {
            retriever.open(ctx, uri)
            val durationMs = retriever.long(MediaMetadataRetriever.METADATA_KEY_DURATION)
            val rotation = retriever.int(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION)
            val rawWidth = retriever.int(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)
            val rawHeight = retriever.int(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)
            // Report what a player would show, not what the encoder wrote.
            val swap = rotation == 90 || rotation == 270
            return ProbeInfo(
                durationMs = durationMs,
                width = if (swap) rawHeight else rawWidth,
                height = if (swap) rawWidth else rawHeight,
                rotation = rotation,
                hasAudio = retriever.string(MediaMetadataRetriever.METADATA_KEY_HAS_AUDIO) == "yes",
                hasVideo = retriever.string(MediaMetadataRetriever.METADATA_KEY_HAS_VIDEO) == "yes",
            )
        } finally {
            retriever.closeQuietly()
        }
    }

    data class ProbeInfo(
        val durationMs: Long,
        val width: Int,
        val height: Int,
        val rotation: Int,
        val hasAudio: Boolean,
        val hasVideo: Boolean,
    ) {
        fun toProbedInput(): ProbedInput = ProbedInput(durationMs, hasAudio, hasVideo)
    }

    /* ---------------------------------------------------------------------------------------- */

    /**
     * One JPEG per requested time, in order. The contract promises the arrays line up, so a time
     * that yields no frame reuses a neighbour rather than shortening the result - a filmstrip with
     * a repeated tile reads as a still moment; one with a missing tile reads as a bug.
     *
     * @param precise cuts the frame at each time rather than the nearest keyframe; see [frameOption]
     *   for what that costs and why it is the caller's decision.
     */
    fun thumbnails(
        ctx: Context,
        uri: String,
        timesMs: List<Long>,
        maxHeight: Int,
        precise: Boolean = false,
    ): List<String> {
        if (timesMs.isEmpty()) return emptyList()
        val height = maxHeight.coerceIn(16, 1080)
        val cacheDir = JobFolders.thumbsCache(ctx).apply { mkdirs() }
        val sourceKey = cacheKey(ctx, uri)

        return synchronized(lockFor(uri)) {
            val files = timesMs.map { File(cacheDir, cacheName(sourceKey, it, height, precise)) }
            // The editor asks for the same times every time it opens, so a strip it has seen before
            // is usually on disk whole - and then opening the source would buy nothing: a
            // descriptor through the content resolver and a container parse in the media server,
            // per clip, on every reopen, for frames nobody reads. The same files come back in the
            // same order the loop below would have handed them. The check is inside the lock
            // because [writeJpeg] writes straight to the final name, and the lock is what keeps a
            // reader from taking a tile that is still being written.
            if (allCached(files)) return@synchronized files.map { Uri.fromFile(it).toString() }
            val retriever = MediaMetadataRetriever()
            val out = ArrayList<String?>(timesMs.size)
            try {
                retriever.open(ctx, uri)
                for ((i, timeMs) in timesMs.withIndex()) {
                    val file = files[i]
                    if (cached(file)) {
                        out += Uri.fromFile(file).toString()
                        continue
                    }
                    val frame = frameAt(retriever, timeMs, height, precise)
                    if (frame == null) {
                        out += null
                        continue
                    }
                    out += try {
                        writeJpeg(frame, file, JPEG_QUALITY_THUMB)
                        Uri.fromFile(file).toString()
                    } catch (e: Exception) {
                        Log.w(TAG, "could not write thumbnail: ${e.message}")
                        null
                    } finally {
                        frame.recycle()
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "thumbnails failed for $uri: ${e.message}")
            } finally {
                retriever.closeQuietly()
            }
            // The contract promises one URI per requested time, in order. If opening the source
            // threw part-way through, the tail is still owed - as placeholders rather than a short
            // array, because a caller lining these up against times would silently mis-align.
            while (out.size < timesMs.size) out += null
            fillGaps(out, cacheDir)
        }
    }

    private fun frameAt(
        retriever: MediaMetadataRetriever,
        timeMs: Long,
        maxHeight: Int,
        precise: Boolean,
    ): Bitmap? {
        val timeUs = timeMs * 1000L
        return try {
            val frame = grab(retriever, timeUs, maxHeight, frameOption(precise))
            // A precise seek is the one that can come back with nothing - past the last frame, or
            // where the container's index is not good enough to decode forward from. The keyframe
            // it would have landed on is a better answer than a gap the caller has to paper over.
            if (frame == null && precise) grab(retriever, timeUs, maxHeight, frameOption(false)) else frame
        } catch (e: Exception) {
            Log.w(TAG, "no frame at ${timeMs}ms: ${e.message}")
            null
        }
    }

    private fun grab(
        retriever: MediaMetadataRetriever,
        timeUs: Long,
        maxHeight: Int,
        option: Int,
    ): Bitmap? =
        if (Build.VERSION.SDK_INT >= 27) {
            // Scales during decode, so a 4K source never materialises a full-size bitmap.
            retriever.getScaledFrameAtTime(timeUs, option, maxHeight * 2, maxHeight)
        } else {
            retriever.getFrameAtTime(timeUs, option)?.let { full ->
                val scale = maxHeight.toFloat() / full.height.toFloat()
                val scaled = Bitmap.createScaledBitmap(
                    full,
                    (full.width * scale).toInt().coerceAtLeast(1),
                    maxHeight,
                    true,
                )
                if (scaled !== full) full.recycle()
                scaled
            }
        }

    /** Whether every tile of a strip is already on disk; an empty file is a write that failed. */
    internal fun allCached(files: List<File>): Boolean = files.all(::cached)

    private fun cached(file: File): Boolean = file.exists() && file.length() > 0L

    /**
     * Which frame a seek settles on.
     *
     * CLOSEST_SYNC is a jump straight to a keyframe and costs one decode, but cameras write a
     * keyframe only every one or two seconds, so several nearby times all land on the same picture -
     * a filmstrip at one frame per second then shows each frame twice over. CLOSEST decodes every
     * frame from that keyframe up to the time asked for instead, so one seek costs as many decodes
     * as there are frames since the last keyframe: about thirty on a 30fps clip with a one-second
     * keyframe interval. Cutting a whole strip precisely therefore costs roughly one decode of the
     * clip, which is why it is the caller's choice and not the default.
     */
    internal fun frameOption(precise: Boolean): Int =
        if (precise) MediaMetadataRetriever.OPTION_CLOSEST else MediaMetadataRetriever.OPTION_CLOSEST_SYNC

    /**
     * Where one frame is cached. Precise frames are kept under their own name because they are a
     * different picture at the same time - a strip asked for precisely must not be served the
     * keyframes a previous request left behind, or the other way round - and the default name is
     * left exactly as it was, so the tiles already on disk are still found.
     */
    internal fun cacheName(sourceKey: String, timeMs: Long, maxHeight: Int, precise: Boolean): String =
        "$sourceKey-$timeMs-$maxHeight${if (precise) "-p" else ""}.jpg"

    /** Replaces nulls with the nearest neighbour, and falls back to a 1x1 black tile. */
    private fun fillGaps(uris: List<String?>, cacheDir: File): List<String> {
        if (uris.all { it != null }) return uris.filterNotNull()
        val filled = uris.toMutableList()
        for (i in filled.indices) {
            if (filled[i] != null) continue
            filled[i] = (i - 1 downTo 0).firstNotNullOfOrNull { filled[it] }
                ?: (i + 1 until filled.size).firstNotNullOfOrNull { filled[it] }
        }
        val placeholder by lazy { Uri.fromFile(blackTile(cacheDir)).toString() }
        return filled.map { it ?: placeholder }
    }

    private fun blackTile(cacheDir: File): File {
        val file = File(cacheDir, "placeholder.jpg")
        if (!file.exists() || file.length() == 0L) {
            val bitmap = Bitmap.createBitmap(2, 2, Bitmap.Config.ARGB_8888)
            bitmap.eraseColor(android.graphics.Color.BLACK)
            try {
                writeJpeg(bitmap, file, JPEG_QUALITY_THUMB)
            } catch (e: Exception) {
                Log.w(TAG, "could not write placeholder tile: ${e.message}")
            } finally {
                bitmap.recycle()
            }
        }
        return file
    }

    /* ---------------------------------------------------------------------------------------- */

    /**
     * The poster frame for a finished render. Uses CLOSEST rather than CLOSEST_SYNC because our own
     * output has one-second keyframes, and half a second off is a visibly different moment.
     */
    fun poster(ctx: Context, videoFile: File, atUs: Long, dest: File): Boolean {
        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(videoFile.absolutePath)
            val bitmap = retriever.getFrameAtTime(atUs, MediaMetadataRetriever.OPTION_CLOSEST)
                ?: retriever.getFrameAtTime(atUs, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
                ?: return false
            try {
                writeJpeg(bitmap, dest, JPEG_QUALITY_POSTER)
                true
            } finally {
                bitmap.recycle()
            }
        } catch (e: Exception) {
            Log.w(TAG, "could not cut poster: ${e.message}")
            false
        } finally {
            retriever.closeQuietly()
        }
    }

    private fun writeJpeg(bitmap: Bitmap, dest: File, quality: Int) {
        dest.parentFile?.mkdirs()
        FileOutputStream(dest).use { out ->
            bitmap.compress(Bitmap.CompressFormat.JPEG, quality, out)
        }
    }

    /* ---------------------------------------------------------------------------------------- */

    /**
     * Identifies the exact bytes behind a URI so an edited or replaced file never serves stale
     * tiles. `content://` has no path to stat, so its size and modification time come from the
     * resolver instead.
     */
    private fun cacheKey(ctx: Context, uri: String): String {
        val parsed = Uri.parse(uri)
        val signature = when (parsed.scheme) {
            "content" -> {
                val (size, modified) = contentStat(ctx, parsed)
                "$uri|$size|$modified"
            }
            else -> {
                val file = parsed.path?.let { File(it) }
                "$uri|${file?.length() ?: 0L}|${file?.lastModified() ?: 0L}"
            }
        }
        val digest = MessageDigest.getInstance("SHA-1").digest(signature.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }.take(24)
    }

    private fun contentStat(ctx: Context, uri: Uri): Pair<Long, Long> = try {
        ctx.contentResolver.query(
            uri,
            arrayOf(OpenableColumns.SIZE, "last_modified"),
            null,
            null,
            null,
        )?.use { cursor ->
            if (cursor.moveToFirst()) {
                val size = if (cursor.columnCount > 0 && !cursor.isNull(0)) cursor.getLong(0) else 0L
                val modified = if (cursor.columnCount > 1 && !cursor.isNull(1)) cursor.getLong(1) else 0L
                size to modified
            } else {
                0L to 0L
            }
        } ?: (0L to 0L)
    } catch (e: Exception) {
        // Plenty of providers do not expose last_modified; the URI itself still distinguishes files.
        0L to 0L
    }

    /* ---------------------------------------------------------------------------------------- */

    private fun MediaMetadataRetriever.open(ctx: Context, uri: String) {
        val parsed = Uri.parse(uri)
        when (parsed.scheme) {
            "content" -> setDataSource(ctx, parsed)
            "file" -> setDataSource(parsed.path ?: uri)
            null -> setDataSource(uri)
            else -> setDataSource(ctx, parsed)
        }
    }

    private fun MediaMetadataRetriever.closeQuietly() {
        try {
            if (Build.VERSION.SDK_INT >= 29) close() else release()
        } catch (e: Exception) {
            Log.w(TAG, "retriever close failed: ${e.message}")
        }
    }

    private fun MediaMetadataRetriever.long(key: Int): Long =
        extractMetadata(key)?.toLongOrNull() ?: 0L

    private fun MediaMetadataRetriever.int(key: Int): Int =
        extractMetadata(key)?.toIntOrNull() ?: 0

    private fun MediaMetadataRetriever.string(key: Int): String? = extractMetadata(key)
}
